{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9847202b-8c65-4cbe-92f2-919cbf10f650",
   "metadata": {},
   "source": [
    "# Purity Reviewer Example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6ed06b3d-93bf-4ab6-8e36-c5dab8251bb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8eba07f4-3c8a-4a23-9ec8-68821ec8aa2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from PurityReviewer.Reviewers.MatchedPurityReviewer import MatchedPurityReviewer\n",
    "from PurityReviewer.Reviewers.ManualPurityReviewer import ManualPurityReviewer\n",
    "from PurityReviewer.AppComponents.utils import download_rdata\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import dalmatian\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19530ec4-dc0d-401f-859f-c22fb8510dd4",
   "metadata": {},
   "source": [
    "# Set up simuated tumor data\n",
    "\n",
    "If `SimulatedTumorData` submodule is not in the `example_notebooks` directory, clone the `SimulatedTumorData` repository in the `example_notebooks` directory\n",
    "```\n",
    "cd example_notebooks\n",
    "git submodule add https://github.com/getzlab/SimulatedTumorData\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0633ffa9-14de-4e12-a554-fc30298eee85",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mWARNING: The directory '/Users/cchu/Library/Caches/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.\u001b[0m\u001b[33m\n",
      "\u001b[0mObtaining file:///Users/cchu/Desktop/Methods/PurityReviewers/example_notebooks/SimulatedTumorData\n",
      "  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: AnnoMate>=0.0.2 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from SimutatedTumorData==0.0.1) (0.0.3)\n",
      "Requirement already satisfied: seaborn in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from SimutatedTumorData==0.0.1) (0.13.0)\n",
      "Requirement already satisfied: dash-cytoscape in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from SimutatedTumorData==0.0.1) (0.3.0)\n",
      "Requirement already satisfied: dash in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from SimutatedTumorData==0.0.1) (2.11.1)\n",
      "Requirement already satisfied: pysam in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from SimutatedTumorData==0.0.1) (0.22.0)\n",
      "Requirement already satisfied: scipy in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from SimutatedTumorData==0.0.1) (1.8.1)\n",
      "Requirement already satisfied: cnv-suite in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.1.1)\n",
      "Requirement already satisfied: dash-bootstrap-components in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.4.1)\n",
      "Requirement already satisfied: dash-daq in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.5.0)\n",
      "Requirement already satisfied: fsspec in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2023.6.0)\n",
      "Requirement already satisfied: gcsfs in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2023.6.0)\n",
      "Requirement already satisfied: google-auth in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.21.0)\n",
      "Requirement already satisfied: google-api-core in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.11.1)\n",
      "Requirement already satisfied: hound in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.0.1)\n",
      "Requirement already satisfied: ipykernel in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (6.22.0)\n",
      "Requirement already satisfied: ipython in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (8.12.0)\n",
      "Requirement already satisfied: jupyterlab>=4.0.2 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (4.0.2)\n",
      "Requirement already satisfied: matplotlib in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (3.5.3)\n",
      "Requirement already satisfied: pandas in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.5.2)\n",
      "Requirement already satisfied: pickleshare in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.7.5)\n",
      "Requirement already satisfied: pillow in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (10.0.0)\n",
      "Requirement already satisfied: pip in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (23.1.2)\n",
      "Requirement already satisfied: plotly>=5.15.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (5.15.0)\n",
      "Requirement already satisfied: setuptools in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (67.8.0)\n",
      "Requirement already satisfied: frozendict in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.3.8)\n",
      "Requirement already satisfied: flask<=2.2.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.2.1)\n",
      "Requirement already satisfied: werkzeug<=2.2.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.2.1)\n",
      "Requirement already satisfied: dash-html-components==2.0.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from dash->SimutatedTumorData==0.0.1) (2.0.0)\n",
      "Requirement already satisfied: dash-core-components==2.0.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from dash->SimutatedTumorData==0.0.1) (2.0.0)\n",
      "Requirement already satisfied: dash-table==5.0.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from dash->SimutatedTumorData==0.0.1) (5.0.0)\n",
      "Requirement already satisfied: typing-extensions>=4.1.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from dash->SimutatedTumorData==0.0.1) (4.6.3)\n",
      "Requirement already satisfied: requests in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from dash->SimutatedTumorData==0.0.1) (2.31.0)\n",
      "Requirement already satisfied: retrying in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from dash->SimutatedTumorData==0.0.1) (1.3.4)\n",
      "Requirement already satisfied: ansi2html in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from dash->SimutatedTumorData==0.0.1) (1.8.0)\n",
      "Requirement already satisfied: nest-asyncio in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from dash->SimutatedTumorData==0.0.1) (1.5.6)\n",
      "Requirement already satisfied: numpy<1.25.0,>=1.17.3 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from scipy->SimutatedTumorData==0.0.1) (1.22.4)\n",
      "Requirement already satisfied: Jinja2>=3.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from flask<=2.2.1->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (3.1.2)\n",
      "Requirement already satisfied: itsdangerous>=2.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from flask<=2.2.1->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.1.2)\n",
      "Requirement already satisfied: click>=8.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from flask<=2.2.1->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (8.1.4)\n",
      "Requirement already satisfied: importlib-metadata>=3.6.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from flask<=2.2.1->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (6.0.0)\n",
      "Requirement already satisfied: async-lru>=1.0.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.0.3)\n",
      "Requirement already satisfied: jupyter-core in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (5.3.0)\n",
      "Requirement already satisfied: jupyter-lsp>=2.0.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.2.0)\n",
      "Requirement already satisfied: jupyter-server<3,>=2.4.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.7.0)\n",
      "Requirement already satisfied: jupyterlab-server<3,>=2.19.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.23.0)\n",
      "Requirement already satisfied: notebook-shim>=0.2 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.2.3)\n",
      "Requirement already satisfied: packaging in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (23.0)\n",
      "Requirement already satisfied: tomli in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.0.1)\n",
      "Requirement already satisfied: tornado>=6.2.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (6.2)\n",
      "Requirement already satisfied: traitlets in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (5.7.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from matplotlib->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.11.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from matplotlib->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (4.40.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from matplotlib->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.4.4)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from matplotlib->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (3.1.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from matplotlib->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from pandas->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2023.3)\n",
      "Requirement already satisfied: tenacity>=6.2.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from plotly>=5.15.0->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (8.2.2)\n",
      "Requirement already satisfied: MarkupSafe>=2.1.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from werkzeug<=2.2.1->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.1.3)\n",
      "Requirement already satisfied: intervaltree<4,>=3.1.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from cnv-suite->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (3.1.0)\n",
      "Requirement already satisfied: natsort<10,>=8.1.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from cnv-suite->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (8.4.0)\n",
      "Requirement already satisfied: pandarallel<1.8,>=1.5.5 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from cnv-suite->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.6.5)\n",
      "Requirement already satisfied: tqdm>=4.45 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from cnv-suite->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (4.65.0)\n",
      "Requirement already satisfied: kaleido<0.3,>=0.2.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from cnv-suite->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.2.1)\n",
      "Requirement already satisfied: nbformat>=4.2.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from cnv-suite->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (5.9.0)\n",
      "Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (3.8.4)\n",
      "Requirement already satisfied: decorator>4.1.2 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (5.1.1)\n",
      "Requirement already satisfied: google-auth-oauthlib in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.0.0)\n",
      "Requirement already satisfied: google-cloud-storage in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.10.0)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from google-auth->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (5.3.1)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from google-auth->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.3.0)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from google-auth->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (4.9)\n",
      "Requirement already satisfied: six>=1.9.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from google-auth->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.16.0)\n",
      "Requirement already satisfied: urllib3<2.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from google-auth->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.26.16)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0.dev0,>=1.56.2 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from google-api-core->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.59.1)\n",
      "Requirement already satisfied: protobuf!=3.20.0,!=3.20.1,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0.dev0,>=3.19.5 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from google-api-core->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (4.23.4)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from requests->dash->SimutatedTumorData==0.0.1) (3.2.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from requests->dash->SimutatedTumorData==0.0.1) (3.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from requests->dash->SimutatedTumorData==0.0.1) (2023.5.7)\n",
      "Requirement already satisfied: appnope in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipykernel->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.1.2)\n",
      "Requirement already satisfied: comm>=0.1.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipykernel->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.1.2)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipykernel->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.6.7)\n",
      "Requirement already satisfied: jupyter-client>=6.1.12 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipykernel->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (8.1.0)\n",
      "Requirement already satisfied: matplotlib-inline>=0.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipykernel->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.1.6)\n",
      "Requirement already satisfied: psutil in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipykernel->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (5.9.0)\n",
      "Requirement already satisfied: pyzmq>=20 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipykernel->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (25.1.0)\n",
      "Requirement already satisfied: backcall in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipython->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.2.0)\n",
      "Requirement already satisfied: jedi>=0.16 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipython->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.18.1)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipython->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (3.0.36)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipython->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.15.1)\n",
      "Requirement already satisfied: stack-data in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipython->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.2.0)\n",
      "Requirement already satisfied: pexpect>4.3 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from ipython->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (4.8.0)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (23.1.0)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (6.0.4)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (4.0.2)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.9.2)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.3.3)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.3.1)\n",
      "Requirement already satisfied: google-cloud-core<3.0dev,>=2.3.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from google-cloud-storage->gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.3.3)\n",
      "Requirement already satisfied: google-resumable-media>=2.3.2 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from google-cloud-storage->gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.5.0)\n",
      "Requirement already satisfied: zipp>=0.5 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from importlib-metadata>=3.6.0->flask<=2.2.1->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (3.11.0)\n",
      "Requirement already satisfied: sortedcontainers<3.0,>=2.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from intervaltree<4,>=3.1.0->cnv-suite->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.4.0)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jedi>=0.16->ipython->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.8.3)\n",
      "Requirement already satisfied: platformdirs>=2.5 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-core->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.5.2)\n",
      "Requirement already satisfied: anyio>=3.1.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (3.7.1)\n",
      "Requirement already satisfied: argon2-cffi in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (21.3.0)\n",
      "Requirement already satisfied: jupyter-events>=0.6.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.6.3)\n",
      "Requirement already satisfied: jupyter-server-terminals in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.4.4)\n",
      "Requirement already satisfied: nbconvert>=6.4.4 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (7.6.0)\n",
      "Requirement already satisfied: overrides in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (7.3.1)\n",
      "Requirement already satisfied: prometheus-client in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.17.0)\n",
      "Requirement already satisfied: send2trash in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.8.2)\n",
      "Requirement already satisfied: terminado>=0.8.3 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.17.1)\n",
      "Requirement already satisfied: websocket-client in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.6.1)\n",
      "Requirement already satisfied: babel>=2.10 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab-server<3,>=2.19.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.12.1)\n",
      "Requirement already satisfied: json5>=0.9.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab-server<3,>=2.19.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.9.14)\n",
      "Requirement already satisfied: jsonschema>=4.17.3 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyterlab-server<3,>=2.19.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (4.18.0)\n",
      "Requirement already satisfied: fastjsonschema in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from nbformat>=4.2.0->cnv-suite->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.17.1)\n",
      "Requirement already satisfied: dill>=0.3.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from pandarallel<1.8,>=1.5.5->cnv-suite->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.3.6)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from pexpect>4.3->ipython->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30->ipython->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.2.5)\n",
      "Requirement already satisfied: pyasn1<0.6.0,>=0.4.6 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from pyasn1-modules>=0.2.1->google-auth->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.5.0)\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from google-auth-oauthlib->gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.3.1)\n",
      "Requirement already satisfied: executing in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from stack-data->ipython->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.8.3)\n",
      "Requirement already satisfied: asttokens in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from stack-data->ipython->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.0.5)\n",
      "Requirement already satisfied: pure-eval in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from stack-data->ipython->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.2.2)\n",
      "Requirement already satisfied: sniffio>=1.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from anyio>=3.1.0->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.3.0)\n",
      "Requirement already satisfied: exceptiongroup in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from anyio>=3.1.0->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.1.2)\n",
      "Requirement already satisfied: google-crc32c<2.0dev,>=1.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from google-resumable-media>=2.3.2->google-cloud-storage->gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.5.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jsonschema>=4.17.3->jupyterlab-server<3,>=2.19.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2023.6.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jsonschema>=4.17.3->jupyterlab-server<3,>=2.19.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.29.1)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jsonschema>=4.17.3->jupyterlab-server<3,>=2.19.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.8.8)\n",
      "Requirement already satisfied: python-json-logger>=2.0.4 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-events>=0.6.0->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.0.7)\n",
      "Requirement already satisfied: pyyaml>=5.3 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-events>=0.6.0->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (6.0)\n",
      "Requirement already satisfied: rfc3339-validator in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-events>=0.6.0->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.1.4)\n",
      "Requirement already satisfied: rfc3986-validator>=0.1.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jupyter-events>=0.6.0->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.1.1)\n",
      "Requirement already satisfied: beautifulsoup4 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (4.12.2)\n",
      "Requirement already satisfied: bleach!=5.0.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (6.0.0)\n",
      "Requirement already satisfied: defusedxml in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.7.1)\n",
      "Requirement already satisfied: jupyterlab-pygments in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.2.2)\n",
      "Requirement already satisfied: mistune<4,>=2.0.3 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (3.0.1)\n",
      "Requirement already satisfied: nbclient>=0.5.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.8.0)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.5.0)\n",
      "Requirement already satisfied: tinycss2 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.2.1)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib->gcsfs->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (3.2.2)\n",
      "Requirement already satisfied: argon2-cffi-bindings in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from argon2-cffi->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (21.2.0)\n",
      "Requirement already satisfied: webencodings in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from bleach!=5.0.0->nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (0.5.1)\n",
      "Requirement already satisfied: fqdn in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jsonschema>=4.17.3->jupyterlab-server<3,>=2.19.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.5.1)\n",
      "Requirement already satisfied: isoduration in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jsonschema>=4.17.3->jupyterlab-server<3,>=2.19.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (20.11.0)\n",
      "Requirement already satisfied: jsonpointer>1.13 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jsonschema>=4.17.3->jupyterlab-server<3,>=2.19.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.4)\n",
      "Requirement already satisfied: uri-template in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jsonschema>=4.17.3->jupyterlab-server<3,>=2.19.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.3.0)\n",
      "Requirement already satisfied: webcolors>=1.11 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from jsonschema>=4.17.3->jupyterlab-server<3,>=2.19.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.13)\n",
      "Requirement already satisfied: cffi>=1.0.1 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from argon2-cffi-bindings->argon2-cffi->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.15.1)\n",
      "Requirement already satisfied: soupsieve>1.2 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from beautifulsoup4->nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.4.1)\n",
      "Requirement already satisfied: pycparser in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi->jupyter-server<3,>=2.4.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (2.21)\n",
      "Requirement already satisfied: arrow>=0.15.0 in /Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages (from isoduration->jsonschema>=4.17.3->jupyterlab-server<3,>=2.19.0->jupyterlab>=4.0.2->AnnoMate>=0.0.2->SimutatedTumorData==0.0.1) (1.2.3)\n",
      "Installing collected packages: SimutatedTumorData\n",
      "  Attempting uninstall: SimutatedTumorData\n",
      "    Found existing installation: SimutatedTumorData 0.0.1\n",
      "    Uninstalling SimutatedTumorData-0.0.1:\n",
      "      Successfully uninstalled SimutatedTumorData-0.0.1\n",
      "  Running setup.py develop for SimutatedTumorData\n",
      "Successfully installed SimutatedTumorData-0.0.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "# Run to install packages for loading the patient and sample data. Only need to run once in your environment.\n",
    "%pip install -e SimulatedTumorData/."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f52bbae6-d4e7-4573-8ff0-ab5e8bd82591",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ff5bef6c-e5bc-41c7-9f24-4bf3991f6af2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from SimulatedTumorData.src.generate_simulated_data import load_patients_and_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d11a1c8d-387c-4ce9-bbd7-3e616dcce87f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/cchu/Desktop/Methods/SimulatedTumorData/sim_data/patient1/phylogicNDT_results_1000\n",
      "loading existing CNV pickle file SimulatedTumorData/sim_data/patient1/patient1.cnv_events.pkl\n",
      "SimulatedTumorData/sim_data/patient1/sample_coverage/p1_t1.binned_coverage.tsv already exists.\n",
      "SimulatedTumorData/sim_data/patient1/sample_coverage/p1_t2.binned_coverage.tsv already exists.\n",
      "SimulatedTumorData/sim_data/patient1/sample_coverage/p1_t3.binned_coverage.tsv already exists.\n",
      "patient variants path exists: SimulatedTumorData/sim_data/patient1/patient1.variants.tsv\n",
      "Sample p1_t1 has variants_fn: SimulatedTumorData/sim_data/patient1/sample_muts/p1_t1.variants.tsv\n",
      "Sample p1_t2 has variants_fn: SimulatedTumorData/sim_data/patient1/sample_muts/p1_t2.variants.tsv\n",
      "Sample p1_t3 has variants_fn: SimulatedTumorData/sim_data/patient1/sample_muts/p1_t3.variants.tsv\n",
      "Run SimulatedTumorData/sim_data/patient1/sample_mut_vcf/p1_t1.variants.vcf through (nexus-snp hg19 RefSeq).\n",
      "Generated sif file: SimulatedTumorData/sim_data/patient1/patient1.phylogicNDT.sif\n",
      "Phylogic results are already loaded here: SimulatedTumorData/sim_data/patient1/phylogicNDT_results_1000\n",
      "Not running commands. Outputs in SimulatedTumorData/sim_data/patient1/sample_ABSOLUTE_results/p1_t1/ABSOLUTE_results\n",
      "Not running commands. Outputs in SimulatedTumorData/sim_data/patient1/sample_ABSOLUTE_results/p1_t2/ABSOLUTE_results\n",
      "Not running commands. Outputs in SimulatedTumorData/sim_data/patient1/sample_ABSOLUTE_results/p1_t3/ABSOLUTE_results\n",
      "Generated cancer_patient_reviewer_sample_data_fn: SimulatedTumorData/sim_data/patient1/cancer_patient_reviewer_data/patient1_samples_data.tsv\n",
      "Generated cancer_patient_reviewer_patient_data_fn: SimulatedTumorData/sim_data/patient1/cancer_patient_reviewer_data/patient1_data.tsv\n",
      "/Users/cchu/Desktop/Methods/SimulatedTumorData/sim_data/patient2/phylogicNDT_results_1000\n",
      "loading existing CNV pickle file SimulatedTumorData/sim_data/patient2/patient2.cnv_events.pkl\n",
      "SimulatedTumorData/sim_data/patient2/sample_coverage/p2_t1.binned_coverage.tsv already exists.\n",
      "SimulatedTumorData/sim_data/patient2/sample_coverage/p2_t2.binned_coverage.tsv already exists.\n",
      "SimulatedTumorData/sim_data/patient2/sample_coverage/p2_t3.binned_coverage.tsv already exists.\n",
      "patient variants path exists: SimulatedTumorData/sim_data/patient2/patient2.variants.tsv\n",
      "Sample p2_t1 has variants_fn: SimulatedTumorData/sim_data/patient2/sample_muts/p2_t1.variants.tsv\n",
      "Sample p2_t2 has variants_fn: SimulatedTumorData/sim_data/patient2/sample_muts/p2_t2.variants.tsv\n",
      "Sample p2_t3 has variants_fn: SimulatedTumorData/sim_data/patient2/sample_muts/p2_t3.variants.tsv\n",
      "Run SimulatedTumorData/sim_data/patient2/sample_mut_vcf/p2_t1.variants.vcf through (nexus-snp hg19 RefSeq).\n",
      "Generated sif file: SimulatedTumorData/sim_data/patient2/patient2.phylogicNDT.sif\n",
      "Phylogic results are already loaded here: SimulatedTumorData/sim_data/patient2/phylogicNDT_results_1000\n",
      "Not running commands. Outputs in SimulatedTumorData/sim_data/patient2/sample_ABSOLUTE_results/p2_t1/ABSOLUTE_results\n",
      "Not running commands. Outputs in SimulatedTumorData/sim_data/patient2/sample_ABSOLUTE_results/p2_t2/ABSOLUTE_results\n",
      "Not running commands. Outputs in SimulatedTumorData/sim_data/patient2/sample_ABSOLUTE_results/p2_t3/ABSOLUTE_results\n",
      "Generated cancer_patient_reviewer_sample_data_fn: SimulatedTumorData/sim_data/patient2/cancer_patient_reviewer_data/patient2_samples_data.tsv\n",
      "Generated cancer_patient_reviewer_patient_data_fn: SimulatedTumorData/sim_data/patient2/cancer_patient_reviewer_data/patient2_data.tsv\n"
     ]
    }
   ],
   "source": [
    "samples, participants = load_patients_and_samples(\n",
    "    path_to_sim_data='SimulatedTumorData/sim_data'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d469372e-3488-4206-8e6c-d23f34158863",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>maf_fn</th>\n",
       "      <th>wxs_purity</th>\n",
       "      <th>collection_date_dfd</th>\n",
       "      <th>cnv_seg_fn</th>\n",
       "      <th>participant_id</th>\n",
       "      <th>preservation_method</th>\n",
       "      <th>wxs_ploidy</th>\n",
       "      <th>ABSOLUTE_pp_calls_tab_fn</th>\n",
       "      <th>ABSOLUTE_pp_modes_data_fn</th>\n",
       "      <th>ABSOLUTE_pp_modes_plots_fn</th>\n",
       "      <th>ABSOLUTE_mode_res_rds_fn</th>\n",
       "      <th>ABSOLUTE_mode_tab_fn</th>\n",
       "      <th>ABSOLUTE_plot_fn</th>\n",
       "      <th>ABSOLUTE_SSNV_mode_res_rds_fn</th>\n",
       "      <th>ABSOLUTE_RData</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p1_t1</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_mu...</td>\n",
       "      <td>0.70</td>\n",
       "      <td>50</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_cn...</td>\n",
       "      <td>patient1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.83</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p1_t2</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_mu...</td>\n",
       "      <td>0.45</td>\n",
       "      <td>100</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_cn...</td>\n",
       "      <td>patient1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.90</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p1_t3</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_mu...</td>\n",
       "      <td>0.90</td>\n",
       "      <td>120</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_cn...</td>\n",
       "      <td>patient1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.75</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2_t1</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_mu...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>100</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_cn...</td>\n",
       "      <td>patient2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.99</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2_t2</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_mu...</td>\n",
       "      <td>0.85</td>\n",
       "      <td>150</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_cn...</td>\n",
       "      <td>patient2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.99</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2_t3</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_mu...</td>\n",
       "      <td>0.20</td>\n",
       "      <td>180</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_cn...</td>\n",
       "      <td>patient2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.00</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      maf_fn  wxs_purity  \\\n",
       "sample_id                                                                  \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_mu...        0.70   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_mu...        0.45   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_mu...        0.90   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_mu...        0.50   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_mu...        0.85   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_mu...        0.20   \n",
       "\n",
       "           collection_date_dfd  \\\n",
       "sample_id                        \n",
       "p1_t1                       50   \n",
       "p1_t2                      100   \n",
       "p1_t3                      120   \n",
       "p2_t1                      100   \n",
       "p2_t2                      150   \n",
       "p2_t3                      180   \n",
       "\n",
       "                                                  cnv_seg_fn participant_id  \\\n",
       "sample_id                                                                     \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_cn...       patient1   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_cn...       patient1   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_cn...       patient1   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_cn...       patient2   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_cn...       patient2   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_cn...       patient2   \n",
       "\n",
       "           preservation_method  wxs_ploidy  \\\n",
       "sample_id                                    \n",
       "p1_t1                      NaN        1.83   \n",
       "p1_t2                      NaN        1.90   \n",
       "p1_t3                      NaN        1.75   \n",
       "p2_t1                      NaN        1.99   \n",
       "p2_t2                      NaN        1.99   \n",
       "p2_t3                      NaN        2.00   \n",
       "\n",
       "                                    ABSOLUTE_pp_calls_tab_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                                   ABSOLUTE_pp_modes_data_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                                  ABSOLUTE_pp_modes_plots_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                                    ABSOLUTE_mode_res_rds_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                                        ABSOLUTE_mode_tab_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                                            ABSOLUTE_plot_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                               ABSOLUTE_SSNV_mode_res_rds_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                                              ABSOLUTE_RData  \n",
       "sample_id                                                     \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...  \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...  \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...  \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...  \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...  \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f2b01e6b-1d5d-439d-8af5-4d268105bb3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>maf_fn</th>\n",
       "      <th>cluster_ccfs_fn</th>\n",
       "      <th>build_tree_posterior_fn</th>\n",
       "      <th>tumor_molecular_subtype</th>\n",
       "      <th>tumor_morphology</th>\n",
       "      <th>tumor_primary_site</th>\n",
       "      <th>cancer_stage</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>death_date_dfd</th>\n",
       "      <th>follow_up_date</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>notes</th>\n",
       "      <th>treatments_fn</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>patient1</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/phylogicN...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/phylogicN...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/phylogicN...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/patient1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient2</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/phylogicN...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/phylogicN...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/phylogicN...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/patient2_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           maf_fn  \\\n",
       "participant_id                                                      \n",
       "patient1        SimulatedTumorData/sim_data/patient1/phylogicN...   \n",
       "patient2        SimulatedTumorData/sim_data/patient2/phylogicN...   \n",
       "\n",
       "                                                  cluster_ccfs_fn  \\\n",
       "participant_id                                                      \n",
       "patient1        SimulatedTumorData/sim_data/patient1/phylogicN...   \n",
       "patient2        SimulatedTumorData/sim_data/patient2/phylogicN...   \n",
       "\n",
       "                                          build_tree_posterior_fn  \\\n",
       "participant_id                                                      \n",
       "patient1        SimulatedTumorData/sim_data/patient1/phylogicN...   \n",
       "patient2        SimulatedTumorData/sim_data/patient2/phylogicN...   \n",
       "\n",
       "               tumor_molecular_subtype tumor_morphology  tumor_primary_site  \\\n",
       "participant_id                                                                \n",
       "patient1                       Unknown          Unknown                 NaN   \n",
       "patient2                       Unknown          Unknown                 NaN   \n",
       "\n",
       "                cancer_stage  vital_status  death_date_dfd  follow_up_date  \\\n",
       "participant_id                                                               \n",
       "patient1                 NaN           NaN             NaN             NaN   \n",
       "patient2                 NaN           NaN             NaN             NaN   \n",
       "\n",
       "                age_at_diagnosis  gender  notes  \\\n",
       "participant_id                                    \n",
       "patient1                      32     NaN    NaN   \n",
       "patient2                      32     NaN    NaN   \n",
       "\n",
       "                                                    treatments_fn  \n",
       "participant_id                                                     \n",
       "patient1        SimulatedTumorData/sim_data/patient1/patient1_...  \n",
       "patient2        SimulatedTumorData/sim_data/patient2/patient2_...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "participants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7363b447-af85-498c-9536-498612c40653",
   "metadata": {},
   "source": [
    "# Reviewer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c8a593ce-2172-4152-9e82-869b3a5e7ff3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from PurityReviewer.AppComponents.utils import parse_absolute_soln, CSIZE_DEFAULT\n",
    "from PurityReviewer.AppComponents.utils import parse_absolute_soln_simulatedTumorData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f043359e-668b-4977-8126-f7196c8ee8b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_pkl_fn = 'simulated_data_purity_review.pkl'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "10b880fd-dd35-4861-92a9-3c44df751ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_reviewer = MatchedPurityReviewer()\n",
    "matched_reviewer.set_review_data(\n",
    "    data_pkl_fn = data_pkl_fn, \n",
    "    description='Matched purity reviewer for simulated data', \n",
    "    df=samples, #pcyc_wm_pairs_df, # optional if directory above already exists. \n",
    "    index=samples.index, #pcyc_wm_pairs_df.index,\n",
    ")\n",
    "matched_reviewer.set_review_app(\n",
    "    sample_info_cols=['ABSOLUTE_plot_fn', 'wxs_purity', 'wxs_ploidy'],\n",
    "    acs_col='cnv_seg_fn', \n",
    "    maf_col='maf_fn',\n",
    "    rdata_fn_col='ABSOLUTE_RData',\n",
    "    mut_fig_hover_data=['Hugo_Symbol', 'Chromosome', 'Start_position'],\n",
    "    csize=CSIZE_DEFAULT,\n",
    "    custom_parse_absolute_soln=parse_absolute_soln_simulatedTumorData # <-- update with my_custom_parse_absolute_soln()\n",
    ")\n",
    "\n",
    "matched_reviewer.set_default_review_data_annotations_configuration()\n",
    "matched_reviewer.set_default_autofill()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2899b183-856c-4159-a46e-bef0587fbd72",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Setting auto_export_path to simulated_data_purity_review.v10.auto_export\n",
      "Making directory simulated_data_purity_review.v10.auto_export for auto exporting.\n",
      "Using simulated_data_purity_review.v10.auto_export for auto exporting.\n",
      "Dash app running on http://0.0.0.0:8099/\n"
     ]
    },
    {
     "data": {
      "application/javascript": [
       "window.open('http://0.0.0.0:8099/')"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "matched_reviewer.run(port=8099, mode='tab', collapsable=False, hide_history_df_cols=['source_data_fn'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62d44268-7544-49c4-9213-b8f65ca32ae5",
   "metadata": {},
   "source": [
    "# Purity Review from CGA characterization pipeline Terra workspace"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc0b43cc-df84-42aa-acd6-95165410d2a3",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dd7b5241-5196-43f5-a830-3f4f7b76dcf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dalmatian"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1daa6617-01c3-41c5-954b-61a459c64d05",
   "metadata": {},
   "source": [
    "```\n",
    "import os\n",
    "os.environ[\"GCLOUD_PROJECT\"] = <google project>\n",
    "\n",
    "wm = dalmatian.WorkspaceManager(<namespace>/<workspace_name>)\n",
    "pairs_df = wm.get_pairs()\n",
    "\n",
    "# Download rdata from ABSOLUTE locally\n",
    "rdata_dir = <directory to save rData>\n",
    "downloaded_rdata_s = download_rdata(pairs_df['absolute_rdata'], rdata_dir=rdata_dir)\n",
    "downloaded_rdata_s.name = 'local_absolute_rdata'\n",
    "pairs_df = pd.concat([pairs_df, downloaded_rdata_s], axis=1)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42a3d9ce-5218-4c78-9f64-ea996595d952",
   "metadata": {},
   "source": [
    "```\n",
    "from PurityReviewer.AppComponents.utils import CSIZE_DEFAULT\n",
    "sex_chr_map = {'23': 'X', '24': 'Y'}\n",
    "rename_chroms = {x: sex_chr_map[x] if x in sex_chr_map.keys() else x for x in CSIZE_DEFAULT.keys()}\n",
    "custom_csize = {f'chr{rename_chroms[chrom]}': length for chrom, length in CSIZE_DEFAULT.items()}\n",
    "custom_csize\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd3e7c16-9a53-4028-b02e-c478d42fc248",
   "metadata": {},
   "source": [
    "Depending on the pipeline or version of ABSOLUTE, the scripts to parse the Rdata may need to be modified. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaeb3159-2dcb-4523-80a5-13c5099823fa",
   "metadata": {},
   "source": [
    "```\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "import rpy2.robjects as robjects\n",
    "\n",
    "def my_custom_parse_absolute_soln(rdata_path: str): # has to be a local path   \n",
    "    absolute_rdata_cols = ['alpha', 'tau', 'tau_hat', '0_line', '1_line',\n",
    "                       'sigma_H', \n",
    "                       'theta_Q', \n",
    "                       'lambda',  \n",
    "                       'SCNA_likelihood', \n",
    "                       'Kar_likelihood', \n",
    "                       'SSNVs_likelihood']\n",
    "    pandas2ri.activate()\n",
    "    print('New version!')\n",
    "    r_list_vector = robjects.r['load'](rdata_path)\n",
    "    r_list_vector = robjects.r[r_list_vector[0]]\n",
    "    # r_data_id = r_list_vector.names[0]\n",
    "\n",
    "    # rdata_tables = r_list_vector.rx2(str(r_data_id))\n",
    "    \n",
    "    # mode_res = rdata_tables.rx2('mode.res')\n",
    "    # mode_tab = mode_res.rx2('mode.tab')\n",
    "    mode_res = r_list_vector.rx2('mode.res')\n",
    "    mode_tab = r_list_vector.rx2('mode.tab') # or segtab?\n",
    "    mod_tab_df = pd.DataFrame(columns=absolute_rdata_cols)\n",
    "    mod_tab_df['alpha'] = mode_tab[:, 0]\n",
    "    mod_tab_df['tau'] = mode_tab[:, 1]\n",
    "    mod_tab_df['tau_hat'] = mode_tab[:, 7]\n",
    "    mod_tab_df['0_line'] = mode_tab[:, 3]\n",
    "    mod_tab_df['step_size'] = mode_tab[:, 4] * 2\n",
    "    mod_tab_df['1_line'] = mod_tab_df['step_size'] + mod_tab_df['0_line']\n",
    "    mod_tab_df['sigma_H'] = mode_tab[:, 8]\n",
    "    mod_tab_df['theta_Q'] = mode_tab[:, 11]\n",
    "    mod_tab_df['lambda'] = mode_tab[:, 12]\n",
    "    mod_tab_df['SCNA_likelihood'] = mode_tab[:, 15]\n",
    "    mod_tab_df['Kar_likelihood'] = mode_tab[:, 17]\n",
    "    mod_tab_df['SSNVs_likelihood'] = mode_tab[:, 20]\n",
    "\n",
    "    return mod_tab_df\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "865be0e0-d49f-4251-b73e-2c2523f19ca6",
   "metadata": {},
   "source": [
    "Run the reviewer. See [CGA WES Characterization Pipeline User Guide\n",
    "](https://docs.google.com/document/d/1VO2kX_fgfUd0x3mBS9NjLUWGZu794WbTepBel3cBg08/edit?usp=sharing) for more inforation about the CGA characterization pipeline outputs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c65b929d-34a7-4daf-9f46-de302397f549",
   "metadata": {},
   "source": [
    "```\n",
    "matched_reviewer = MatchedPurityReviewer()\n",
    "matched_reviewer.set_review_data(\n",
    "    data_pkl_fn = <pkl path>, \n",
    "    description= <description>, \n",
    "    df=pairs_df,\n",
    "    index=pairs_df.index,\n",
    ")\n",
    "matched_reviewer.set_review_app(\n",
    "    sample_info_cols=['participant', 'alleliccapseg_plot'],\n",
    "    acs_col='alleliccapseg_tsv', \n",
    "    maf_col='mutation_validator_validated_maf',\n",
    "    rdata_fn_col='local_absolute_rdata',\n",
    "    mut_fig_hover_data=['Hugo_Symbol', 'Chromosome', 'Start_position'],\n",
    "    csize=CSIZE_DEFAULT,\n",
    "    custom_parse_absolute_soln=parse_absolute_soln # <-- update with my_custom_parse_absolute_soln()\n",
    ")\n",
    "\n",
    "matched_reviewer.set_default_review_data_annotations_configuration()\n",
    "matched_reviewer.set_default_autofill()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ce32779-9760-4888-8f19-146a2eb16d94",
   "metadata": {},
   "source": [
    "```\n",
    "matched_reviewer.run(port=8099, review_data_table_df=luad_pairs_df[['Purity', 'participant']], mode='tab')\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:purity_reviewer_annomate_env]",
   "language": "python",
   "name": "conda-env-purity_reviewer_annomate_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
