{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9847202b-8c65-4cbe-92f2-919cbf10f650",
   "metadata": {},
   "source": [
    "# Purity Reviewer Example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6ed06b3d-93bf-4ab6-8e36-c5dab8251bb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8eba07f4-3c8a-4a23-9ec8-68821ec8aa2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from PurityReviewer.Reviewers.MatchedPurityReviewer import MatchedPurityReviewer\n",
    "from PurityReviewer.Reviewers.ManualPurityReviewer import ManualPurityReviewer\n",
    "from PurityReviewer.AppComponents.utils import download_rdata\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import dalmatian\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19530ec4-dc0d-401f-859f-c22fb8510dd4",
   "metadata": {},
   "source": [
    "# Set up simuated tumor data\n",
    "\n",
    "Clone the `SimulatedTumorData` repo and install in your environment.\n",
    "```\n",
    "cd <path to put SimulatedTumorData repo>\n",
    "git clone https://github.com/getzlab/SimulatedTumorData\n",
    "\n",
    "conda activate <your env>\n",
    "pip install -e <path to put SimulatedTumorData repo>\n",
    "```\n",
    "\n",
    "This only needs to be done once."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f52bbae6-d4e7-4573-8ff0-ab5e8bd82591",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ff5bef6c-e5bc-41c7-9f24-4bf3991f6af2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from SimulatedTumorData.src.generate_simulated_data import load_patients_and_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d11a1c8d-387c-4ce9-bbd7-3e616dcce87f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/cchu/Desktop/Methods/SimulatedTumorData/sim_data/patient1/phylogicNDT_results_1000\n",
      "loading existing CNV pickle file SimulatedTumorData/sim_data/patient1/patient1.cnv_events.pkl\n",
      "SimulatedTumorData/sim_data/patient1/sample_coverage/p1_t1.binned_coverage.tsv already exists.\n",
      "SimulatedTumorData/sim_data/patient1/sample_coverage/p1_t2.binned_coverage.tsv already exists.\n",
      "SimulatedTumorData/sim_data/patient1/sample_coverage/p1_t3.binned_coverage.tsv already exists.\n",
      "patient variants path exists: SimulatedTumorData/sim_data/patient1/patient1.variants.tsv\n",
      "Sample p1_t1 has variants_fn: SimulatedTumorData/sim_data/patient1/sample_muts/p1_t1.variants.tsv\n",
      "Sample p1_t2 has variants_fn: SimulatedTumorData/sim_data/patient1/sample_muts/p1_t2.variants.tsv\n",
      "Sample p1_t3 has variants_fn: SimulatedTumorData/sim_data/patient1/sample_muts/p1_t3.variants.tsv\n",
      "Run SimulatedTumorData/sim_data/patient1/sample_mut_vcf/p1_t1.variants.vcf through (nexus-snp hg19 RefSeq).\n",
      "Generated sif file: SimulatedTumorData/sim_data/patient1/patient1.phylogicNDT.sif\n",
      "Phylogic results are already loaded here: SimulatedTumorData/sim_data/patient1/phylogicNDT_results_1000\n",
      "Not running commands. Outputs in SimulatedTumorData/sim_data/patient1/sample_ABSOLUTE_results/p1_t1/ABSOLUTE_results\n",
      "Not running commands. Outputs in SimulatedTumorData/sim_data/patient1/sample_ABSOLUTE_results/p1_t2/ABSOLUTE_results\n",
      "Not running commands. Outputs in SimulatedTumorData/sim_data/patient1/sample_ABSOLUTE_results/p1_t3/ABSOLUTE_results\n",
      "Generated cancer_patient_reviewer_sample_data_fn: SimulatedTumorData/sim_data/patient1/cancer_patient_reviewer_data/patient1_samples_data.tsv\n",
      "Generated cancer_patient_reviewer_patient_data_fn: SimulatedTumorData/sim_data/patient1/cancer_patient_reviewer_data/patient1_data.tsv\n",
      "/Users/cchu/Desktop/Methods/SimulatedTumorData/sim_data/patient2/phylogicNDT_results_1000\n",
      "loading existing CNV pickle file SimulatedTumorData/sim_data/patient2/patient2.cnv_events.pkl\n",
      "SimulatedTumorData/sim_data/patient2/sample_coverage/p2_t1.binned_coverage.tsv already exists.\n",
      "SimulatedTumorData/sim_data/patient2/sample_coverage/p2_t2.binned_coverage.tsv already exists.\n",
      "SimulatedTumorData/sim_data/patient2/sample_coverage/p2_t3.binned_coverage.tsv already exists.\n",
      "patient variants path exists: SimulatedTumorData/sim_data/patient2/patient2.variants.tsv\n",
      "Sample p2_t1 has variants_fn: SimulatedTumorData/sim_data/patient2/sample_muts/p2_t1.variants.tsv\n",
      "Sample p2_t2 has variants_fn: SimulatedTumorData/sim_data/patient2/sample_muts/p2_t2.variants.tsv\n",
      "Sample p2_t3 has variants_fn: SimulatedTumorData/sim_data/patient2/sample_muts/p2_t3.variants.tsv\n",
      "Run SimulatedTumorData/sim_data/patient2/sample_mut_vcf/p2_t1.variants.vcf through (nexus-snp hg19 RefSeq).\n",
      "Generated sif file: SimulatedTumorData/sim_data/patient2/patient2.phylogicNDT.sif\n",
      "Phylogic results are already loaded here: SimulatedTumorData/sim_data/patient2/phylogicNDT_results_1000\n",
      "Not running commands. Outputs in SimulatedTumorData/sim_data/patient2/sample_ABSOLUTE_results/p2_t1/ABSOLUTE_results\n",
      "Not running commands. Outputs in SimulatedTumorData/sim_data/patient2/sample_ABSOLUTE_results/p2_t2/ABSOLUTE_results\n",
      "Not running commands. Outputs in SimulatedTumorData/sim_data/patient2/sample_ABSOLUTE_results/p2_t3/ABSOLUTE_results\n",
      "Generated cancer_patient_reviewer_sample_data_fn: SimulatedTumorData/sim_data/patient2/cancer_patient_reviewer_data/patient2_samples_data.tsv\n",
      "Generated cancer_patient_reviewer_patient_data_fn: SimulatedTumorData/sim_data/patient2/cancer_patient_reviewer_data/patient2_data.tsv\n"
     ]
    }
   ],
   "source": [
    "samples, participants = load_patients_and_samples(\n",
    "    path_to_sim_data='SimulatedTumorData/sim_data'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d469372e-3488-4206-8e6c-d23f34158863",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>maf_fn</th>\n",
       "      <th>wxs_purity</th>\n",
       "      <th>collection_date_dfd</th>\n",
       "      <th>cnv_seg_fn</th>\n",
       "      <th>participant_id</th>\n",
       "      <th>preservation_method</th>\n",
       "      <th>wxs_ploidy</th>\n",
       "      <th>ABSOLUTE_pp_calls_tab_fn</th>\n",
       "      <th>ABSOLUTE_pp_modes_data_fn</th>\n",
       "      <th>ABSOLUTE_pp_modes_plots_fn</th>\n",
       "      <th>ABSOLUTE_mode_res_rds_fn</th>\n",
       "      <th>ABSOLUTE_mode_tab_fn</th>\n",
       "      <th>ABSOLUTE_plot_fn</th>\n",
       "      <th>ABSOLUTE_SSNV_mode_res_rds_fn</th>\n",
       "      <th>ABSOLUTE_RData</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p1_t1</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_mu...</td>\n",
       "      <td>0.70</td>\n",
       "      <td>50</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_cn...</td>\n",
       "      <td>patient1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.83</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p1_t2</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_mu...</td>\n",
       "      <td>0.45</td>\n",
       "      <td>100</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_cn...</td>\n",
       "      <td>patient1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.90</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p1_t3</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_mu...</td>\n",
       "      <td>0.90</td>\n",
       "      <td>120</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_cn...</td>\n",
       "      <td>patient1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.75</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/sample_AB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2_t1</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_mu...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>100</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_cn...</td>\n",
       "      <td>patient2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.99</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2_t2</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_mu...</td>\n",
       "      <td>0.85</td>\n",
       "      <td>150</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_cn...</td>\n",
       "      <td>patient2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.99</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2_t3</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_mu...</td>\n",
       "      <td>0.20</td>\n",
       "      <td>180</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_cn...</td>\n",
       "      <td>patient2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.00</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/sample_AB...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      maf_fn  wxs_purity  \\\n",
       "sample_id                                                                  \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_mu...        0.70   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_mu...        0.45   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_mu...        0.90   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_mu...        0.50   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_mu...        0.85   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_mu...        0.20   \n",
       "\n",
       "           collection_date_dfd  \\\n",
       "sample_id                        \n",
       "p1_t1                       50   \n",
       "p1_t2                      100   \n",
       "p1_t3                      120   \n",
       "p2_t1                      100   \n",
       "p2_t2                      150   \n",
       "p2_t3                      180   \n",
       "\n",
       "                                                  cnv_seg_fn participant_id  \\\n",
       "sample_id                                                                     \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_cn...       patient1   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_cn...       patient1   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_cn...       patient1   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_cn...       patient2   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_cn...       patient2   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_cn...       patient2   \n",
       "\n",
       "           preservation_method  wxs_ploidy  \\\n",
       "sample_id                                    \n",
       "p1_t1                      NaN        1.83   \n",
       "p1_t2                      NaN        1.90   \n",
       "p1_t3                      NaN        1.75   \n",
       "p2_t1                      NaN        1.99   \n",
       "p2_t2                      NaN        1.99   \n",
       "p2_t3                      NaN        2.00   \n",
       "\n",
       "                                    ABSOLUTE_pp_calls_tab_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                                   ABSOLUTE_pp_modes_data_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                                  ABSOLUTE_pp_modes_plots_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                                    ABSOLUTE_mode_res_rds_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                                        ABSOLUTE_mode_tab_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                                            ABSOLUTE_plot_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                               ABSOLUTE_SSNV_mode_res_rds_fn  \\\n",
       "sample_id                                                      \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...   \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...   \n",
       "\n",
       "                                              ABSOLUTE_RData  \n",
       "sample_id                                                     \n",
       "p1_t1      SimulatedTumorData/sim_data/patient1/sample_AB...  \n",
       "p1_t2      SimulatedTumorData/sim_data/patient1/sample_AB...  \n",
       "p1_t3      SimulatedTumorData/sim_data/patient1/sample_AB...  \n",
       "p2_t1      SimulatedTumorData/sim_data/patient2/sample_AB...  \n",
       "p2_t2      SimulatedTumorData/sim_data/patient2/sample_AB...  \n",
       "p2_t3      SimulatedTumorData/sim_data/patient2/sample_AB...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f2b01e6b-1d5d-439d-8af5-4d268105bb3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>maf_fn</th>\n",
       "      <th>cluster_ccfs_fn</th>\n",
       "      <th>build_tree_posterior_fn</th>\n",
       "      <th>tumor_molecular_subtype</th>\n",
       "      <th>tumor_morphology</th>\n",
       "      <th>tumor_primary_site</th>\n",
       "      <th>cancer_stage</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>death_date_dfd</th>\n",
       "      <th>follow_up_date</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>notes</th>\n",
       "      <th>treatments_fn</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>participant_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>patient1</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/phylogicN...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/phylogicN...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/phylogicN...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient1/patient1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient2</th>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/phylogicN...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/phylogicN...</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/phylogicN...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SimulatedTumorData/sim_data/patient2/patient2_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           maf_fn  \\\n",
       "participant_id                                                      \n",
       "patient1        SimulatedTumorData/sim_data/patient1/phylogicN...   \n",
       "patient2        SimulatedTumorData/sim_data/patient2/phylogicN...   \n",
       "\n",
       "                                                  cluster_ccfs_fn  \\\n",
       "participant_id                                                      \n",
       "patient1        SimulatedTumorData/sim_data/patient1/phylogicN...   \n",
       "patient2        SimulatedTumorData/sim_data/patient2/phylogicN...   \n",
       "\n",
       "                                          build_tree_posterior_fn  \\\n",
       "participant_id                                                      \n",
       "patient1        SimulatedTumorData/sim_data/patient1/phylogicN...   \n",
       "patient2        SimulatedTumorData/sim_data/patient2/phylogicN...   \n",
       "\n",
       "               tumor_molecular_subtype tumor_morphology  tumor_primary_site  \\\n",
       "participant_id                                                                \n",
       "patient1                       Unknown          Unknown                 NaN   \n",
       "patient2                       Unknown          Unknown                 NaN   \n",
       "\n",
       "                cancer_stage  vital_status  death_date_dfd  follow_up_date  \\\n",
       "participant_id                                                               \n",
       "patient1                 NaN           NaN             NaN             NaN   \n",
       "patient2                 NaN           NaN             NaN             NaN   \n",
       "\n",
       "                age_at_diagnosis  gender  notes  \\\n",
       "participant_id                                    \n",
       "patient1                      32     NaN    NaN   \n",
       "patient2                      32     NaN    NaN   \n",
       "\n",
       "                                                    treatments_fn  \n",
       "participant_id                                                     \n",
       "patient1        SimulatedTumorData/sim_data/patient1/patient1_...  \n",
       "patient2        SimulatedTumorData/sim_data/patient2/patient2_...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "participants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7363b447-af85-498c-9536-498612c40653",
   "metadata": {},
   "source": [
    "# Reviewer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c8a593ce-2172-4152-9e82-869b3a5e7ff3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from PurityReviewer.AppComponents.utils import parse_absolute_soln, CSIZE_DEFAULT\n",
    "from PurityReviewer.AppComponents.utils import parse_absolute_soln_simulatedTumorData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f043359e-668b-4977-8126-f7196c8ee8b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = 'simulated_data_purity_review'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "10b880fd-dd35-4861-92a9-3c44df751ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_reviewer = MatchedPurityReviewer()\n",
    "matched_reviewer.set_review_data(\n",
    "    data_path=data_path, \n",
    "    description='Matched purity reviewer for simulated data', \n",
    "    df=samples,\n",
    "    index=samples.index, \n",
    ")\n",
    "matched_reviewer.set_review_app(\n",
    "    sample_info_cols=['ABSOLUTE_plot_fn', 'wxs_purity', 'wxs_ploidy'],\n",
    "    acs_col='cnv_seg_fn', \n",
    "    maf_col='maf_fn',\n",
    "    rdata_fn_col='ABSOLUTE_RData',\n",
    "    mut_fig_hover_data=['Hugo_Symbol', 'Chromosome', 'Start_position'],\n",
    "    csize=CSIZE_DEFAULT,\n",
    "    custom_parse_absolute_soln=parse_absolute_soln_simulatedTumorData # <-- update with my_custom_parse_absolute_soln()\n",
    ")\n",
    "\n",
    "matched_reviewer.set_default_review_data_annotations_configuration()\n",
    "matched_reviewer.set_default_autofill()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2899b183-856c-4159-a46e-bef0587fbd72",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Setting auto_export_path to simulated_data_purity_review/data.auto_export\n",
      "Using simulated_data_purity_review/data.auto_export for auto exporting.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages/AnnoMate/ReviewDataApp.py:738: UserWarning:\n",
      "\n",
      "You are in test mode. Your data will not be saved.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash app running on http://0.0.0.0:8099/\n"
     ]
    },
    {
     "data": {
      "application/javascript": [
       "window.open('http://0.0.0.0:8099/')"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "matched_reviewer.run(port=8099, mode='tab', collapsable=False, hide_history_df_cols=['source_data_fn'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e04b008f-8990-461f-99ab-bc32f3ea621c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/cchu/opt/anaconda3/envs/purity_reviewer_annomate_env/lib/python3.9/site-packages/AnnoMate/ReviewerTemplate.py:477: UserWarning:\n",
      "\n",
      "Data is not frozen. Annotations will not be saved. Please freeze data in the dashboard to save annotations.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Purity</th>\n",
       "      <th>Ploidy</th>\n",
       "      <th>Method</th>\n",
       "      <th>Absolute_solution_idx</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>value</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p1_t1</th>\n",
       "      <td>0.70</td>\n",
       "      <td>1.8</td>\n",
       "      <td>Absolute</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p1_t2</th>\n",
       "      <td>0.45</td>\n",
       "      <td>1.8</td>\n",
       "      <td>Absolute</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p1_t3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2_t1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2_t2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p2_t3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Purity  Ploidy    Method  Absolute_solution_idx Notes\n",
       "value                                                       \n",
       "p1_t1    0.70     1.8  Absolute                    1.0      \n",
       "p1_t2    0.45     1.8  Absolute                    1.0      \n",
       "p1_t3     NaN     NaN                              NaN      \n",
       "p2_t1     NaN     NaN                              NaN      \n",
       "p2_t2     NaN     NaN                              NaN      \n",
       "p2_t3     NaN     NaN                              NaN      "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_reviewer.get_annot()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62d44268-7544-49c4-9213-b8f65ca32ae5",
   "metadata": {},
   "source": [
    "# Purity Review from CGA characterization pipeline Terra workspace"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc0b43cc-df84-42aa-acd6-95165410d2a3",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dd7b5241-5196-43f5-a830-3f4f7b76dcf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dalmatian"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1daa6617-01c3-41c5-954b-61a459c64d05",
   "metadata": {},
   "source": [
    "```\n",
    "import os\n",
    "os.environ[\"GCLOUD_PROJECT\"] = <google project>\n",
    "\n",
    "wm = dalmatian.WorkspaceManager(<namespace>/<workspace_name>)\n",
    "pairs_df = wm.get_pairs()\n",
    "\n",
    "# Download rdata from ABSOLUTE locally\n",
    "rdata_dir = <directory to save rData>\n",
    "downloaded_rdata_s = download_rdata(pairs_df['absolute_rdata'], rdata_dir=rdata_dir)\n",
    "downloaded_rdata_s.name = 'local_absolute_rdata'\n",
    "pairs_df = pd.concat([pairs_df, downloaded_rdata_s], axis=1)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42a3d9ce-5218-4c78-9f64-ea996595d952",
   "metadata": {},
   "source": [
    "```\n",
    "from PurityReviewer.AppComponents.utils import CSIZE_DEFAULT\n",
    "sex_chr_map = {'23': 'X', '24': 'Y'}\n",
    "rename_chroms = {x: sex_chr_map[x] if x in sex_chr_map.keys() else x for x in CSIZE_DEFAULT.keys()}\n",
    "custom_csize = {f'chr{rename_chroms[chrom]}': length for chrom, length in CSIZE_DEFAULT.items()}\n",
    "custom_csize\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd3e7c16-9a53-4028-b02e-c478d42fc248",
   "metadata": {},
   "source": [
    "Depending on the pipeline or version of ABSOLUTE, the scripts to parse the Rdata may need to be modified. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaeb3159-2dcb-4523-80a5-13c5099823fa",
   "metadata": {},
   "source": [
    "```\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "import rpy2.robjects as robjects\n",
    "\n",
    "def my_custom_parse_absolute_soln(rdata_path: str): # has to be a local path   \n",
    "    absolute_rdata_cols = ['alpha', 'tau', 'tau_hat', '0_line', '1_line',\n",
    "                       'sigma_H', \n",
    "                       'theta_Q', \n",
    "                       'lambda',  \n",
    "                       'SCNA_likelihood', \n",
    "                       'Kar_likelihood', \n",
    "                       'SSNVs_likelihood']\n",
    "    pandas2ri.activate()\n",
    "    print('New version!')\n",
    "    r_list_vector = robjects.r['load'](rdata_path)\n",
    "    r_list_vector = robjects.r[r_list_vector[0]]\n",
    "    # r_data_id = r_list_vector.names[0]\n",
    "\n",
    "    # rdata_tables = r_list_vector.rx2(str(r_data_id))\n",
    "    \n",
    "    # mode_res = rdata_tables.rx2('mode.res')\n",
    "    # mode_tab = mode_res.rx2('mode.tab')\n",
    "    mode_res = r_list_vector.rx2('mode.res')\n",
    "    mode_tab = r_list_vector.rx2('mode.tab') # or segtab?\n",
    "    mod_tab_df = pd.DataFrame(columns=absolute_rdata_cols)\n",
    "    mod_tab_df['alpha'] = mode_tab[:, 0]\n",
    "    mod_tab_df['tau'] = mode_tab[:, 1]\n",
    "    mod_tab_df['tau_hat'] = mode_tab[:, 7]\n",
    "    mod_tab_df['0_line'] = mode_tab[:, 3]\n",
    "    mod_tab_df['step_size'] = mode_tab[:, 4] * 2\n",
    "    mod_tab_df['1_line'] = mod_tab_df['step_size'] + mod_tab_df['0_line']\n",
    "    mod_tab_df['sigma_H'] = mode_tab[:, 8]\n",
    "    mod_tab_df['theta_Q'] = mode_tab[:, 11]\n",
    "    mod_tab_df['lambda'] = mode_tab[:, 12]\n",
    "    mod_tab_df['SCNA_likelihood'] = mode_tab[:, 15]\n",
    "    mod_tab_df['Kar_likelihood'] = mode_tab[:, 17]\n",
    "    mod_tab_df['SSNVs_likelihood'] = mode_tab[:, 20]\n",
    "\n",
    "    return mod_tab_df\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "865be0e0-d49f-4251-b73e-2c2523f19ca6",
   "metadata": {},
   "source": [
    "Run the reviewer. See [CGA WES Characterization Pipeline User Guide\n",
    "](https://docs.google.com/document/d/1VO2kX_fgfUd0x3mBS9NjLUWGZu794WbTepBel3cBg08/edit?usp=sharing) for more inforation about the CGA characterization pipeline outputs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c65b929d-34a7-4daf-9f46-de302397f549",
   "metadata": {},
   "source": [
    "```\n",
    "matched_reviewer = MatchedPurityReviewer()\n",
    "matched_reviewer.set_review_data(\n",
    "    data_path = <data path>, \n",
    "    description= <description>, \n",
    "    df=pairs_df,\n",
    "    index=pairs_df.index,\n",
    ")\n",
    "matched_reviewer.set_review_app(\n",
    "    sample_info_cols=['participant', 'alleliccapseg_plot'],\n",
    "    acs_col='alleliccapseg_tsv', \n",
    "    maf_col='mutation_validator_validated_maf',\n",
    "    rdata_fn_col='local_absolute_rdata',\n",
    "    mut_fig_hover_data=['Hugo_Symbol', 'Chromosome', 'Start_position'],\n",
    "    csize=CSIZE_DEFAULT,\n",
    "    custom_parse_absolute_soln=parse_absolute_soln # <-- update with my_custom_parse_absolute_soln()\n",
    ")\n",
    "\n",
    "matched_reviewer.set_default_review_data_annotations_configuration()\n",
    "matched_reviewer.set_default_autofill()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ce32779-9760-4888-8f19-146a2eb16d94",
   "metadata": {},
   "source": [
    "```\n",
    "matched_reviewer.run(port=8099, review_data_table_df=luad_pairs_df[['Purity', 'participant']], mode='tab')\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:purity_reviewer_annomate_env]",
   "language": "python",
   "name": "conda-env-purity_reviewer_annomate_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
